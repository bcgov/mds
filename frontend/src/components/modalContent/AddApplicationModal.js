import React from "react";
import PropTypes from "prop-types";
import AddApplicationForm from "@/components/Forms/applications/AddApplicationForm";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  title: PropTypes.string,
};

const defaultProps = {
  title: "",
};

export const AddApplicationModal = (props) => (
  <div>
    <AddApplicationForm {...props} />
  </div>
);

AddApplicationModal.propTypes = propTypes;
AddApplicationModal.defaultProps = defaultProps;

export default AddApplicationModal;
