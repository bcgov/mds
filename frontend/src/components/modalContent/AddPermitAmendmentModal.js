import React from "react";
import PropTypes from "prop-types";
import AddPermitAmendmentForm from "@/components/Forms/AddPermitAmendmentForm";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  title: PropTypes.string,
};

const defaultProps = {
  title: "",
};

export const AddPermitAmendmentModal = (props) => (
  <div>
    <AddPermitAmendmentForm {...props} />
  </div>
);

AddPermitAmendmentModal.propTypes = propTypes;
AddPermitAmendmentModal.defaultProps = defaultProps;

export default AddPermitAmendmentModal;
