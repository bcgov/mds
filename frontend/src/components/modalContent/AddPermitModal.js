import React from "react";
import PropTypes from "prop-types";
import AddPermitForm from "@/components/Forms/AddPermitForm";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  title: PropTypes.string,
};

const defaultProps = {
  title: "",
};

export const AddPermitModal = (props) => (
  <div>
    <AddPermitForm {...props} />
  </div>
);

AddPermitModal.propTypes = propTypes;
AddPermitModal.defaultProps = defaultProps;

export default AddPermitModal;
