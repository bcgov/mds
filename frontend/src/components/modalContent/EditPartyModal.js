import React from "react";
import PropTypes from "prop-types";
import EditFullPartyForm from "@/components/Forms/parties/EditFullPartyForm";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  title: PropTypes.string,
};

const defaultProps = {
  title: "",
};

export const EditPartyModal = (props) => (
  <div>
    <EditFullPartyForm {...props} />
  </div>
);

EditPartyModal.propTypes = propTypes;
EditPartyModal.defaultProps = defaultProps;
export default EditPartyModal;
