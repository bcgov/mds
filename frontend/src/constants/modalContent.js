export const ADD_MINE_MANAGER = "Add Mine Manager";
export const UPDATE_MINE_MANAGER = "Update Mine Manager";
export const UPDATE_PERMITTEE = "Update Permittee";
export const ADD_TENURE = "Add Tenure Number";
export const CREATE_MINE_RECORD = "Create Mine Record";
export const UPDATE_MINE_RECORD = "Update Mine Record";
export const ADD_TAILINGS = "+ Add TSF";
export const EDIT_TAILINGS_REPORT = "Update TSF Report";
export const ADD_TAILINGS_REPORT = "Add TSF Report";
export const ADD_CONTACT = "Add New Contact";
export const EDIT_PARTY = (partyName) => `Update ${partyName}`;

export const PERSON = "PER";
export const ORGANIZATION = "ORG";

export const PARTY_NOT_FOUND = `If the party is not in the list above, please add a new person or company below`;
export const PERSON_NOT_FOUND =
  "If the party is not in the list above, please add a new person below";
export const COMPANY_NOT_FOUND =
  "If the company is not in the list above, please add a new company below";
export const EDIT_PARTY_RELATIONSHIP = "Edit Contact";

export const ADD_PERMIT = "Add a new permit";
export const PERMIT_AMENDMENT = "Add new amendment";
export const EDIT_PERMIT = "Edit Permit";

export const ADD_APPLICATION = "Add Application";
export const Edit_APPLICATION = "Edit Application";
