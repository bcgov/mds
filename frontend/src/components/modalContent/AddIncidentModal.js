// Passing props into function causes linter to not recognize use of props used in that function.
/* eslint-disable react/no-unused-prop-types */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { getFormValues } from "redux-form";
import { concat, reject } from "lodash";
import { Steps, Button, Popconfirm } from "antd";
import * as FORM from "@/constants/forms";
import AddIncidentReportingForm from "@/components/Forms/incidents/AddIncidentReportingForm";
import AddIncidentDetailForm from "@/components/Forms/incidents/AddIncidentDetailForm";
import AddIncidentFollowUpForm from "@/components/Forms/incidents/AddIncidentFollowUpForm";

import CustomPropTypes from "@/customPropTypes";

const { Step } = Steps;

const propTypes = {
  newIncident: PropTypes.bool.isRequired,
  onSubmit: PropTypes.func.isRequired,
  afterClose: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  incidentDeterminationOptions: CustomPropTypes.options.isRequired,
  incidentStatusCodeOptions: CustomPropTypes.options.isRequired,
  followupActionOptions: CustomPropTypes.options.isRequired,
  doSubparagraphOptions: CustomPropTypes.options.isRequired,
  initialValues: PropTypes.objectOf(PropTypes.any).isRequired,
  inspectors: CustomPropTypes.options.isRequired,
  addIncidentFormValues: PropTypes.objectOf(PropTypes.any),
  mineGuid: PropTypes.string.isRequired,
};

const defaultProps = {
  addIncidentFormValues: {},
};

const invalidReportingPayload = (values) =>
  !(
    values.reported_timestamp &&
    values.reported_by_name &&
    values.reported_to_inspector_party_guid &&
    values.responsible_inspector_party_guid
  );

const invalidDetailPayload = (values) =>
  !(
    values.incident_timestamp &&
    values.incident_description &&
    values.determination_type_code &&
    // If DO, need subparagraphs
    ((values.determination_type_code === "DO" &&
      values.determination_inspector_party_guid &&
      values.dangerous_occurrence_subparagraph_ids &&
      values.dangerous_occurrence_subparagraph_ids.length !== 0) ||
      (values.determination_type_code === "NDO" &&
        values.status_code &&
        values.determination_inspector_party_guid) ||
      values.determination_type_code === "PEN")
  );

const invalidFollowUpPayload = (values) =>
  !(values.status_code && values.followup_investigation_type_code);

const actionVerb = (newIncident) => {
  if (newIncident) return <span>Save&nbsp;</span>;
  return <span>Edit&nbsp;</span>;
};

const StepForms = (
  props,
  next,
  prev,
  handleIncidentSubmit,
  uploadedFiles,
  onFileLoad,
  onRemoveFile
) => [
  {
    title: "Initial Report",
    content: (
      <AddIncidentReportingForm initialValues={props.initialValues} inspectors={props.inspectors} />
    ),
    buttons: (
      <Button
        id="step1-next"
        type="tertiary"
        className="full-mobile"
        onClick={() => next()}
        disabled={invalidReportingPayload(props.addIncidentFormValues)}
      >
        Next
      </Button>
    ),
  },
  {
    title: "Add Details",
    content: (
      <AddIncidentDetailForm
        mineGuid={props.mineGuid}
        initialValues={props.initialValues}
        doSubparagraphOptions={props.doSubparagraphOptions}
        incidentDeterminationOptions={props.incidentDeterminationOptions}
        incidentStatusCodeOptions={props.incidentStatusCodeOptions}
        inspectors={props.inspectors}
        doDetermination={props.addIncidentFormValues.determination_type_code}
        uploadedFiles={uploadedFiles.filter(
          (file) => file.mine_incident_document_type_code === "INI"
        )}
        onFileLoad={onFileLoad}
        onRemoveFile={onRemoveFile}
      />
    ),
    buttons: (
      <span>
        <Button id="step-back" type="tertiary" className="full-mobile" onClick={() => prev()}>
          Back
        </Button>
        {props.addIncidentFormValues.determination_type_code !== "NDO" && (
          <Button
            id="step2-next"
            type="tertiary"
            className="full-mobile"
            onClick={() => next()}
            disabled={invalidDetailPayload(props.addIncidentFormValues)}
          >
            Next
          </Button>
        )}
        <Button
          type="primary"
          className="full-mobile"
          onClick={(event) => handleIncidentSubmit(event, false)}
          disabled={invalidDetailPayload(props.addIncidentFormValues)}
        >
          {actionVerb(props.newIncident)}
          {props.addIncidentFormValues.determination_type_code !== "NDO" && (
            <span>Initial&nbsp;</span>
          )}
          Incident
        </Button>
      </span>
    ),
  },
  {
    title: "Follow Up",
    content: (
      <AddIncidentFollowUpForm
        mineGuid={props.mineGuid}
        initialValues={props.initialValues}
        incidentDeterminationOptions={props.incidentDeterminationOptions}
        followupActionOptions={props.followupActionOptions}
        incidentStatusCodeOptions={props.incidentStatusCodeOptions}
        hasFatalities={props.addIncidentFormValues.number_of_fatalities > 0 || false}
        hasFollowUp={props.addIncidentFormValues.followup_inspection || false}
        uploadedFiles={uploadedFiles.filter(
          (file) => file.mine_incident_document_type_code === "FIN"
        )}
        onFileLoad={onFileLoad}
        onRemoveFile={onRemoveFile}
      />
    ),
    buttons: [
      <Button id="step-back" type="tertiary" className="full-mobile" onClick={() => prev()}>
        Back
      </Button>,
      <Button
        type="primary"
        className="full-mobile"
        onClick={(event) => handleIncidentSubmit(event, false)}
        disabled={invalidFollowUpPayload(props.addIncidentFormValues)}
      >
        {actionVerb(props.newIncident)}
        Incident
      </Button>,
    ],
  },
];

export class AddIncidentModal extends Component {
  state = {
    current: 0,
    uploadedFiles: this.props.initialValues.documents
      ? [...this.props.initialValues.documents]
      : [],
  };

  handleIncidentSubmit = () => {
    this.props.onSubmit({
      ...this.props.addIncidentFormValues,
      updated_documents: this.state.uploadedFiles,
    });
    // TODO: Catch error
    this.close();
  };

  close = () => {
    this.props.closeModal();
    this.props.afterClose();
  };

  next = () => this.setState((prevState) => ({ current: prevState.current + 1 }));

  prev = () => this.setState((prevState) => ({ current: prevState.current - 1 }));

  onFileLoad = (document_name, document_manager_guid, mine_incident_document_type_code) =>
    this.setState((prevState) => ({
      uploadedFiles: concat(prevState.uploadedFiles, {
        document_name,
        document_manager_guid,
        mine_incident_document_type_code,
      }),
    }));

  onRemoveFile = (file) => {
    this.setState((prevState) => ({
      uploadedFiles: reject(
        prevState.uploadedFiles,
        (uploadedFile) => file.document_manager_guid === uploadedFile.document_manager_guid
      ),
    }));
  };

  render = () => {
    const Forms = StepForms(
      this.props,
      this.next,
      this.prev,
      this.handleIncidentSubmit,
      this.state.uploadedFiles,
      this.onFileLoad,
      this.onRemoveFile
    );

    return (
      <div>
        <div>
          <div>
            <Steps current={this.state.current}>
              {Forms.map((step) => (
                <Step key={step.title} title={step.title} />
              ))}
            </Steps>
            <br />

            <div>{Forms[this.state.current].content}</div>

            <div className="right center-mobile">
              <Popconfirm
                placement="top"
                title="Are you sure you want to cancel?"
                okText="Yes"
                cancelText="No"
                onConfirm={this.close}
              >
                <Button type="secondary" className="full-mobile">
                  Cancel
                </Button>
              </Popconfirm>

              {Forms[this.state.current].buttons}
            </div>
          </div>
        </div>
      </div>
    );
  };
}

const mapStateToProps = (state) => ({
  addIncidentFormValues: getFormValues(FORM.MINE_INCIDENT)(state) || {},
});

const mapDispatchToProps = (dispatch) => bindActionCreators({}, dispatch);
AddIncidentModal.propTypes = propTypes;
AddIncidentModal.defaultProps = defaultProps;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddIncidentModal);
