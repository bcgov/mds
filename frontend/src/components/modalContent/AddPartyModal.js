import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { getFormValues, submit, reset, change } from "redux-form";
import { Row, Col, Steps, Button, Popconfirm } from "antd";
import * as FORM from "@/constants/forms";
import CustomPropTypes from "@/customPropTypes";
import { createParty, addPartyRelationship } from "@/actionCreators/partiesActionCreator";
import { fetchMineNameList } from "@/actionCreators/mineActionCreator";
import { getMineNames } from "@/selectors/mineSelectors";
import AddFullPartyForm from "@/components/Forms/parties/AddFullPartyForm";
import AddRolesForm from "@/components/Forms/parties/AddRolesForm";

const { Step } = Steps;

const propTypes = {
  fetchData: PropTypes.func.isRequired,
  fetchMineNameList: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  createParty: PropTypes.func.isRequired,
  change: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
  addPartyFormValues: PropTypes.objectOf(PropTypes.strings),
  addRolesFormValues: PropTypes.objectOf(PropTypes.strings),
  addPartyForm: PropTypes.objectOf(CustomPropTypes.genericFormState),
  provinceOptions: PropTypes.arrayOf(CustomPropTypes.dropdownListItem).isRequired,
  addPartyRelationship: PropTypes.func.isRequired,
  partyRelationshipTypesList: PropTypes.arrayOf(CustomPropTypes.dropdownListItem).isRequired,
  mineNameList: PropTypes.arrayOf(CustomPropTypes.mineName).isRequired,
};

const defaultProps = {
  addPartyFormValues: {},
  addRolesFormValues: {},
  addPartyForm: {},
};

const groupRolePayloads = (formValues, party_guid) => {
  const rolePayloads = {};
  Object.entries(formValues).forEach(([key, value]) => {
    const [field, roleNumber] = key.split("-");
    rolePayloads[roleNumber] = rolePayloads[roleNumber] || {};
    rolePayloads[roleNumber][field] = value;
    rolePayloads[roleNumber].party_guid = party_guid;
  });
  return rolePayloads;
};

const invalidPartyPayload = (addPartyFormValues, isPerson) =>
  (isPerson && addPartyFormValues.first_name === undefined) ||
  addPartyFormValues.party_name === undefined ||
  addPartyFormValues.phone_no === undefined;

const invalidRolePayload = (roleNumbers, addRolesFormValues) =>
  Object.keys(addRolesFormValues).length > 0 &&
  roleNumbers.some(
    (roleNumber) =>
      addRolesFormValues[`mine_guid-${roleNumber}`] === undefined ||
      addRolesFormValues[`mine_party_appt_type_code-${roleNumber}`] === undefined
  );

export class AddPartyModal extends Component {
  state = { isPerson: true, current: 0, roleNumbers: [] };

  componentWillMount() {
    this.props.fetchMineNameList();
  }

  handleChange = (roleNumber) => (name) => {
    if (name.length > 2) {
      this.props.fetchMineNameList({ name });
    } else if (name.length === 0) {
      this.props.change(FORM.ADD_ROLES, `mine_guid-${roleNumber}`, name);
      this.props.fetchMineNameList();
    }
  };

  handleSelect = (roleNumber) => (value) => {
    this.props.change(FORM.ADD_ROLES, `mine_guid-${roleNumber}`, value);
  };

  togglePartyChange = (value) => {
    this.setState({ isPerson: value.target.value });
  };

  handlePartySubmit = async (event, addAnother) => {
    event.preventDefault();
    const type = this.state.isPerson ? "PER" : "ORG";
    const payload = { type, ...this.props.addPartyFormValues };
    const party = await this.props
      .createParty(payload)
      .then(({ data }) => {
        this.props.reset(FORM.ADD_FULL_PARTY);
        if (addAnother) {
          this.prev();
        } else {
          this.props.fetchData();
          this.props.closeModal();
        }
        return data;
      })
      .catch(() => {
        this.prev();
      });

    if (!party) {
      return Promise.resolve();
    }

    const rolePayloads = groupRolePayloads(this.props.addRolesFormValues, party.party_guid);
    const createdRoles = Object.values(rolePayloads).map(this.props.addPartyRelationship);

    return Promise.all(createdRoles).then(() => {
      this.props.reset(FORM.ADD_ROLES);
    });
  };

  clearFieldValues = (roleNumber) => {
    this.props.change(FORM.ADD_ROLES, `mine_guid-${roleNumber}`, "");
    this.props.change(FORM.ADD_ROLES, `mine_party_appt_type_code-${roleNumber}`, "");
    this.props.change(FORM.ADD_ROLES, `start_date-${roleNumber}`, "");
    this.props.change(FORM.ADD_ROLES, `end_date-${roleNumber}`, "");
  };

  addField = () => {
    this.setState(({ roleNumbers: prevNumbers }) => {
      const highestRoleNumber = Number(prevNumbers[prevNumbers.length - 1] || 0);
      const newRoleNumber = String(highestRoleNumber + 1);
      return { roleNumbers: [...prevNumbers, newRoleNumber], activeKey: newRoleNumber };
    });
  };

  removeField = (roleNumber) => () => {
    // Clear field values from Redux store
    this.clearFieldValues(roleNumber);
    // Remove role number from state
    this.setState(({ roleNumbers: prevNumbers }) => ({
      roleNumbers: prevNumbers.filter((x) => x !== roleNumber),
    }));
  };

  cancel = () => {
    this.props.closeModal();
    this.props.reset(FORM.ADD_FULL_PARTY);
    this.props.reset(FORM.ADD_ROLES);
  };

  handleActivePanelChange = (roleNumber) => {
    this.setState({ activeKey: roleNumber });
  };

  next() {
    this.setState((prevState) => ({ current: prevState.current + 1 }));
  }

  prev() {
    this.setState((prevState) => ({ current: prevState.current - 1 }));
  }

  renderStepOne() {
    return (
      <AddFullPartyForm
        isPerson={this.state.isPerson}
        togglePartyChange={this.togglePartyChange}
        provinceOptions={this.props.provinceOptions}
      />
    );
  }

  // WIP
  renderStepTwo() {
    return (
      <div>
        <Row gutter={48}>
          <Col
            md={12}
            sm={24}
            xs={24}
            className="border--right--layout"
            style={{ marginTop: "16px" }}
          >
            <p>
              You cannot add a role of Permittee or Engineer of Record through this section. Please
              go to the designated mine.
            </p>
            <AddRolesForm
              roleNumbers={this.state.roleNumbers}
              addField={this.addField}
              removeField={this.removeField}
              partyRelationshipTypesList={this.props.partyRelationshipTypesList}
              mineNameList={this.props.mineNameList}
              handleChange={this.handleChange}
              handleSelect={this.handleSelect}
              activeKey={this.state.activeKey}
              handleActivePanelChange={this.handleActivePanelChange}
            />
          </Col>

          <Col md={12} sm={24} xs={24} style={{ marginTop: "16px" }}>
            <p>
              If you would like to add another contact, click on the button below. Your current
              contact will be submitted once you opt to add a new contact.
            </p>
            <Button
              type="primary"
              className="full-mobile"
              style={{ marginLeft: 0 }}
              onClick={(event) => this.handlePartySubmit(event, true)}
            >
              Submit and Add another contact
            </Button>
          </Col>
        </Row>
      </div>
    );
  }

  render() {
    const steps = [
      {
        title: "Basic Information",
        content: this.renderStepOne(),
      },
      {
        title: "Role",
        content: this.renderStepTwo(),
      },
    ];

    return (
      <div>
        <div>
          <Steps current={this.state.current}>
            {steps.map((step) => (
              <Step key={step.title} title={step.title} />
            ))}
          </Steps>
          <div>{steps[this.state.current].content}</div>
          <div className="right center-mobile">
            <Popconfirm
              placement="top"
              title="Are you sure you want to cancel?"
              okText="Yes"
              cancelText="No"
              onConfirm={this.cancel}
            >
              <Button type="secondary" className="full-mobile">
                Cancel
              </Button>
            </Popconfirm>
            {this.state.current > 0 && (
              <Button type="tertiary" className="full-mobile" onClick={() => this.prev()}>
                Previous
              </Button>
            )}
            {this.state.current < steps.length - 1 && (
              <Button
                type="primary"
                className="full-mobile"
                onClick={() => this.next()}
                disabled={invalidPartyPayload(this.props.addPartyFormValues, this.state.isPerson)}
              >
                Next
              </Button>
            )}
            {this.state.current === steps.length - 1 && (
              <Button
                type="primary"
                className="full-mobile"
                onClick={(event) => this.handlePartySubmit(event, false)}
                disabled={invalidRolePayload(this.state.roleNumbers, this.props.addRolesFormValues)}
              >
                Submit
              </Button>
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  addPartyFormValues: getFormValues(FORM.ADD_FULL_PARTY)(state) || {},
  addRolesFormValues: getFormValues(FORM.ADD_ROLES)(state) || {},
  mineNameList: getMineNames(state).mines,
  addPartyForm: state.form[FORM.ADD_FULL_PARTY],
});
const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      addPartyRelationship,
      fetchMineNameList,
      submit,
      reset,
      createParty,
      change,
    },
    dispatch
  );

AddPartyModal.propTypes = propTypes;
AddPartyModal.defaultProps = defaultProps;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddPartyModal);
