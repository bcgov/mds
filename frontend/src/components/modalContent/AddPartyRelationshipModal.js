import React, { Component } from "react";
import { Radio, Button, Row, Col } from "antd";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import PropTypes from "prop-types";
import AddPartyRelationshipForm from "@/components/Forms/PartyRelationships/AddPartyRelationshipForm";
import AddQuickPartyForm from "@/components/Forms/parties/AddQuickPartyForm";
import * as ModalContent from "@/constants/modalContent";
import { fetchParties, createParty } from "@/actionCreators/partiesActionCreator";
import { getRawParties } from "@/selectors/partiesSelectors";
import { SUCCESS_CHECKMARK } from "@/constants/assets";
import CustomPropTypes from "@/customPropTypes";
import { createItemMap, createItemIdsArray } from "@/utils/helpers";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  fetchParties: PropTypes.func.isRequired,
  createParty: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  partyRelationshipType: CustomPropTypes.partyRelationshipType.isRequired,
  parties: PropTypes.arrayOf(CustomPropTypes.party),
  mine: CustomPropTypes.mine.isRequired,
};

const defaultProps = {
  parties: [],
};

export class AddPartyRelationshipModal extends Component {
  state = { isPerson: true, isFormVisible: false, successfulPartyCreation: false };

  componentDidMount() {
    this.setState({ isPerson: this.props.partyRelationshipType.person });
  }

  togglePartyChange = (value) => {
    this.setState({ isPerson: value.target.value });
  };

  toggleFormVisibility = () => {
    this.setState((prevState) => ({ isFormVisible: !prevState.isFormVisible }));
  };

  handlePartySubmit = (values) => {
    const type = this.state.isPerson ? "PER" : "ORG";
    const payload = { type, ...values };
    this.props
      .createParty(payload)
      .then(() => {
        this.toggleFormVisibility();
        this.props.fetchParties();
        this.setState({ successfulPartyCreation: true });
      })
      .catch();
  };

  renderRadioButtonGroup = (person, organization) =>
    person &&
    organization && (
      <Radio.Group defaultValue size="large" onChange={this.togglePartyChange}>
        <Radio.Button value>Person</Radio.Button>
        <Radio.Button value={false}>Company</Radio.Button>
      </Radio.Group>
    );

  render() {
    let filteredParties = this.props.parties;
    if (!this.props.partyRelationshipType.person) {
      filteredParties = filteredParties.filter(({ party_type_code }) => party_type_code === "ORG");
    } else if (!this.props.partyRelationshipType.organization) {
      filteredParties = filteredParties.filter(({ party_type_code }) => party_type_code === "PER");
    }

    return (
      <div>
        <Row gutter={48}>
          <Col md={12} sm={24}>
            <AddPartyRelationshipForm
              onSubmit={this.props.onSubmit}
              handleChange={this.props.handleChange}
              closeModal={this.props.closeModal}
              title={this.props.title}
              partyRelationshipType={this.props.partyRelationshipType}
              parties={createItemMap(filteredParties, "party_guid")}
              partyIds={createItemIdsArray(filteredParties, "party_guid")}
              mine={this.props.mine}
            />
          </Col>
          <Col md={12} sm={24}>
            {!this.state.isFormVisible && (
              <div className="center">
                {!this.state.successfulPartyCreation && (
                  <div style={{ marginTop: "80px" }}>
                    <p>
                      {this.props.partyRelationshipType.person &&
                        this.props.partyRelationshipType.organization &&
                        ModalContent.PARTY_NOT_FOUND}
                      {!this.props.partyRelationshipType.person &&
                        this.props.partyRelationshipType.organization &&
                        ModalContent.COMPANY_NOT_FOUND}
                      {this.props.partyRelationshipType.person &&
                        !this.props.partyRelationshipType.organization &&
                        ModalContent.PERSON_NOT_FOUND}
                    </p>
                    <Button type="secondary" onClick={this.toggleFormVisibility}>
                      Add new Contact
                    </Button>
                  </div>
                )}
                {this.state.successfulPartyCreation && (
                  <div>
                    <img src={SUCCESS_CHECKMARK} alt="success" />
                    <h4>Success, your contact was created! </h4>
                    <p>
                      Please go back to the name section and input your new contact into the desired
                      role.
                    </p>
                  </div>
                )}
              </div>
            )}
            {this.state.isFormVisible && (
              <div className="center">
                {this.renderRadioButtonGroup(
                  this.props.partyRelationshipType.person,
                  this.props.partyRelationshipType.organization
                )}
                <AddQuickPartyForm
                  onSubmit={this.handlePartySubmit}
                  isPerson={this.state.isPerson}
                />
              </div>
            )}
          </Col>
        </Row>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  parties: getRawParties(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchParties,
      createParty,
    },
    dispatch
  );

AddPartyRelationshipModal.propTypes = propTypes;
AddPartyRelationshipModal.defaultProps = defaultProps;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddPartyRelationshipModal);
