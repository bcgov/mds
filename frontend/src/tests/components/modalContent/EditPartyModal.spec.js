import React from "react";
import { shallow } from "enzyme";
import { EditPartyModal } from "@/components/modalContent/EditPartyModal";

const dispatchProps = {};
const props = {};

const setupDispatchProps = () => {
  dispatchProps.oSubmit = jest.fn();
};

const setupProps = () => {
  props.title = "mockTitle";
};

beforeEach(() => {
  setupDispatchProps();
  setupProps();
});

describe("EditPartyModal", () => {
  it("renders properly", () => {
    const component = shallow(<EditPartyModal {...dispatchProps} {...props} />);
    expect(component).toMatchSnapshot();
  });
});
