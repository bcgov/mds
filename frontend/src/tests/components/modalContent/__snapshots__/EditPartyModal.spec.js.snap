// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPartyModal renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditPartyModal
    closeModal={[MockFunction]}
    isPerson={true}
    oSubmit={[MockFunction]}
    provinceOptions={
      Array [
        Object {
          "label": "BC",
          "value": "BC",
        },
        Object {
          "label": "AB",
          "value": "AB",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ReduxForm
        closeModal={[MockFunction]}
        initialValues={undefined}
        isPerson={true}
        onSubmit={undefined}
        provinceOptions={
          Array [
            Object {
              "label": "BC",
              "value": "BC",
            },
            Object {
              "label": "AB",
              "value": "AB",
            },
          ]
        }
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "closeModal": [MockFunction],
        "initialValues": undefined,
        "isPerson": true,
        "onSubmit": undefined,
        "provinceOptions": Array [
          Object {
            "label": "BC",
            "value": "BC",
          },
          Object {
            "label": "AB",
            "value": "AB",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ReduxForm
          closeModal={[MockFunction]}
          initialValues={undefined}
          isPerson={true}
          onSubmit={undefined}
          provinceOptions={
            Array [
              Object {
                "label": "BC",
                "value": "BC",
              },
              Object {
                "label": "AB",
                "value": "AB",
              },
            ]
          }
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "closeModal": [MockFunction],
          "initialValues": undefined,
          "isPerson": true,
          "onSubmit": undefined,
          "provinceOptions": Array [
            Object {
              "label": "BC",
              "value": "BC",
            },
            Object {
              "label": "AB",
              "value": "AB",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
